test/e2e/cli_test.go

```diff
@@ -1733,6 +1735,138 @@ func (s *CLISuite) TestPluginStruct() {
+    workflows.argoproj.io/version: '>= 3.5.0'
+spec:
+  entrypoint: main
+  templates:
+    - name: main
+      steps:
+        - - name: build
+            template: output-artifact
+            hooks:
+              started:
+                expression: steps["build"].status == "Running"
+                template: started
+              success:
+                expression: steps["build"].status == "Succeeded"
+                template: success
+              failed:
+                expression: steps["build"].status == "Failed" || steps["build"].status == "Error"
+                template: failed
+        - - name: print
+            template: print-artifact
+            arguments:
+              artifacts:
+                - name: message
+                  from: "{{steps.build.outputs.artifacts.result}}"
+    
+    - name: output-artifact
+      script:
+        image: python:alpine3.6
+        command: [ python ]
+        source: |
+          import time
+          import random
+          import sys
+          time.sleep(random.randint(0,5)) # lifecycle hook for running won't trigger unless it runs for more than "a few seconds"
+          with open("result.txt", "w") as f:
+            f.write("Welcome")
+          if {{retries}} == 2:
+          	sys.exit(0)
+          sys.exit(1)
+      retryStrategy: 
+        limit: 2
+      outputs:
+        artifacts:
+          - name: result
+            path: /result.txt
+
+    - name: started
+      container:
+        image: python:alpine3.6
+        command: [sh, -c]
+        args: ["echo STARTED!"]
+
+    - name: success
+      container:
+        image: python:alpine3.6
+        command: [sh, -c]
+        args: ["echo SUCCEEDED!"]
+
+    - name: failed
+      container:
+        image: python:alpine3.6
+        command: [sh, -c]
+        args: ["echo FAILED or ERROR!"]
+
+    - name: print-artifact
+      inputs:
+        artifacts:
+          - name: message
+            path: /tmp/message
+      container:
+        image: python:alpine3.6
+        command: [sh, -c]
+        args: ["cat /tmp/message"]
+`).When().
+		SubmitWorkflow().
+		WaitForWorkflow(fixtures.ToBeCompleted).
+		Then().
+		ExpectWorkflow(func(t *testing.T, metadata *metav1.ObjectMeta, status *wfv1.WorkflowStatus) {
+			workflowName = metadata.Name
+			assert.True(t, status.Fulfilled())
+			assert.Equal(t, wfv1.WorkflowSucceeded, status.Phase)
+			for _, node := range status.Nodes {
+				if node.Type == wfv1.NodeTypeRetry {
+					assert.Equal(t, wfv1.NodeSucceeded, node.Phase)
+					children = node.Children
+				}
+			}
+		}).
+		ExpectWorkflowNode(func(status wfv1.NodeStatus) bool {
+			return status.Name == "retries-with-hooks-and-artifact[0].build(0)"
+		}, func(t *testing.T, status *wfv1.NodeStatus, pod *corev1.Pod) {
+			assert.Contains(t, children, status.ID)
+			assert.Equal(t, false, status.NodeFlag.Hooked)
+		}).
+		ExpectWorkflowNode(func(status wfv1.NodeStatus) bool {
```

Comment:
I think we should add test for other hooks as well, not limited to `hooks.started`